### PR TITLE
Added Rational Curves

### PR DESCRIPTION
**Pull request by IQuick143**
# Objective

- Fix NURBS curves not being NURBS curves

## Solution

- Implement `RationalCurve` and `RationalSegment` and use them to implement NURBS curves.

## Changelog

Added `RationalCurve` and `RationalSegment`, made NURBS curves a RationalGenerator, added a `From<CubicCurve> for RationalCurve` impl and added a test for it.

### CODE REVIEW

File: crates/bevy_math/src/cubic_splines.rs
Comment (JohnTheCoolingFan) [inline crates/bevy_math/src/cubic_splines.rs]: I think you forgot to change these strings
```suggestion
        let cubic_velocities: Vec<_> = b_spline.iter_velocities(10).collect();
        let rational_velocities: Vec<_> = rational_b_spline.iter_velocities(10).collect();
        compare_vectors(cubic_velocities, rational_velocities, "velocity");

        let cubic_accelerations: Vec<_> = b_spline.iter_accelerations(10).collect();
        let rational_accelerations: Vec<_> = rational_b_spline.iter_accelerations(10).collect();
        compare_vectors(cubic_accelerations, rational_accelerations, "acceleration");
```

**Review by JohnTheCoolingFan — CHANGES_REQUESTED** [review]

**Review by JohnTheCoolingFan — APPROVED** [review]

**Comment by JohnTheCoolingFan** [comment]
I would like to wait for @NthTensor to review as well before merging

**Comment by NthTensor** [comment]
Looking it over now.

File: crates/bevy_math/src/cubic_splines.rs
Comment (NthTensor) [inline crates/bevy_math/src/cubic_splines.rs:1071]: Segments should be determined by spans in the knot vector.

Each successive pair of knots k_i, k_i+1 represents the interval [k_i, k_i+1) on which the input t maps to piecewise segment i. In the event than k_i = k_i+1, the span is zero and we skip segment i.

We have rationality, but we still have to get non-uniformity right.

File: crates/bevy_math/src/cubic_splines.rs
Comment (NthTensor) [inline crates/bevy_math/src/cubic_splines.rs:870]: We should probably also add a `span: Range<f32>` here derived from the knot value, then use it to look up the appropriate segment.

File: crates/bevy_math/src/cubic_splines.rs
Comment (NthTensor) [inline crates/bevy_math/src/cubic_splines.rs:837]: In some cases, as when the knot vector has multiplicity at the ends, we may actually need to use the same control point multiple times.

File: crates/bevy_math/src/cubic_splines.rs
Comment (NthTensor) [inline crates/bevy_math/src/cubic_splines.rs:430]: I _think_ we also need to normalize the knot vector so that it covers the interval [0, n], because this is the domain of our parameterized curve.

**Review by NthTensor — APPROVED** [review]
Ok this is a tremendous, fantastic improvement but there are still seem to be issues remaining with the knot vector, segmentation, and parameterization.

I think the best option would be to merge this first, then do the fixes. I can PR this myself, but if either of you prefer to tackle this I'm happy to provide a more comprehensive description of the problem.

File: crates/bevy_math/src/cubic_splines.rs
Comment (NthTensor) [inline crates/bevy_math/src/cubic_splines.rs:1177]: This is a really good test. Ideally, I would like at-least one more using comparing points on a nonuniform, weighted, 3d curve against "ground-truth" data computed using tinyspline (with appropriate deltas to accommodate floating point error).

Comment (IQuick143) [reply crates/bevy_math/src/cubic_splines.rs:1177]: Yeah I agree a ground-truth test is also needed, but I felt that was over the scope of what I was doing.

Comment (IQuick143) [reply crates/bevy_math/src/cubic_splines.rs:1177]: You can either add it yourself when this gets merged, or if you have a test-case in mind, send me the data and I'll put it in.

Comment (NthTensor) [reply crates/bevy_math/src/cubic_splines.rs:1177]: Yep, just noting what I'm sure Alice will want. We should totally merge this into the PR first.

**Comment by JohnTheCoolingFan** [comment]
Awesome review @NthTensor. Tell me when you're done with this guys and I'll merge and we can continue with other things.

File: crates/bevy_math/src/cubic_splines.rs
Comment (IQuick143) [reply crates/bevy_math/src/cubic_splines.rs:1071]: Ah, I thought the matrix already encoded that information.

**Pull request merged**
